nixos/modules/services/web-apps/invidious.nix

```diff
@@ -0,0 +1,156 @@
+{ lib, config, pkgs, ... }:
+let
+  cfg = config.services.invidious;
+
+  # Default from https://github.com/kemalcr/kemal/blob/15022c25b824ecbfe465862bfa8a8e37a4b6ab40/src/kemal/config.cr#L31
+  port = cfg.settings.external_port or 3000;
+
+  databaseUser = "kemal";
+  databaseName = "invidious";
+
+  # TODO: Automatic db cleanup as described in https://github.com/omarroth/invidious/wiki/Database-Information-and-Maintenance
+  # TODO: Add a test
+  # TODO: Detect rollbacks, warn for potential problems
+
+  workingDir = pkgs.runCommandNoCC "invidious-config" {
+    config = builtins.toJSON cfg.settings;
+    passAsFile = "config";
+  } ''
+    mkdir -p $out/config
+    mv "$configPath" $out/config/config.yml
+  '';
+
+in {
+
+  options.services.invidious = {
+    enable = lib.mkEnableOption "Invidious";
+
+    settings =
+      let jsonType = with lib.types; nullOr (oneOf
+        [ bool str int (listOf jsonType) (attrsOf jsonType) ] );
```

Comment:
I didn't know this was possibile. How does a recursive type shows up in the manual?

Reply:
Good point, it doesn't, because it gives infinite recursion first :). I added a commit to fix this by overriding the description.